### ISSUE TITLE
Native notification support for Windows

### ISSUE THREAD

**Issue by Aetherinox** [xdpirate/ntfy-electron#2]
Thus far, the app is great.

Only question is if this app has the ability to send push notifications to the host computer (Windows).

I closed out the browser window which was how I received notifications to Windows, and this app seems to only show them if I open the app and view the interface. No notifications are ever pushed to the Windows notification system.

I read someone else asking about this as well, so I figured I'd ask to see if it's possible / already integrated.

**Comment by xdpirate** [comment]
Yes, this used to work on all platforms, before ntfy changed its implementation of native notifications to use service workers. I'm not quite sure how to fix it, but I will have a closer look at it soon.

**Comment by Aetherinox** [comment]
I took some time after my post, and decided to go in and write a new integration for notifications, as well as compatibility with macos, linux, and windows.

If you want, I can PR it after. I had to have a whole bunch of features added, including subscribing to particular topics, support for api token calls, etc.

Not done yet, but thus far notifications have been working well. I use them every day.

**Comment by xdpirate** [comment]
Please do submit a PR, I would be very interested in this! I want to keep it as "native" to the web app as possible though, and subscribing to topics etc are already features there. It is a wrapper for the web app, after all. But please, submit a PR and I will review it :)

**Comment by Aetherinox** [comment]
Yeah mine was just adding a way to do it via the GUI (laziness maybe). But as far as my needs, I didn't need anything on the electron side edited, wanted to keep that original so that it's still the webapp.

Aside from notifications, the remaining were just "nice to have"

- Notifications
- Flag so that exit button actually minimizes it to tray instead of closing. 
  - I don't remember what app got me started with that, but when I initially installed, I was hitting the close button to minimize and it was killing the app. So I added a flag to force it to close to tray as an alternative (with right-click context menu option to quit).
- Hotkeys
- API token, Topics
  - Used `custom-electron-prompt` to keep conformity, didn't want to start adding a bunch of new crap or manually doing a bunch of UIs

Other than that, really didn't want to start transforming the app. Just notifications and a few optional quality of life features.

Biggest things were notifications and hitting that damn close button and killing it. I must've done that 30 times on the first day. But I don't want to divert too much from that. Thus far, it's worked great.

And I can also set hotkeys to be enabled with a cli argument, that way the functionality doesn't change and start screwing with people.

**Comment by xdpirate** [comment]
>Flag so that exit button actually minimizes it to tray instead of closing. 

This was already on my "to do when I get around to it" mental list, so that's great. All looks good to me, looking forward to the PR :) Feel free to add info to the readme about the added features as well.

**Comment by Aetherinox** [comment]
Sounds good. Once I finish up I'll write up some notes. Other than that stuff, I can't really think of anything else the app needs. 99.9% of it runs through ntfy, so that should keep control.

**Comment by Aetherinox** [comment]
Side note just to document it.

Upgraded electron just for security, all worked well, however, there's an issue between `custom-electron-prompt v1.5.7` and `electron v31.x`.

If you open a custom prompt dialog and you press the close button on the top right, it flat out crashes the app and it hangs in the background. You can only close the dialog using "OK" or "Cancel".

So I backed it up to electron v30 and it works fine. I haven't done a deep dive into if it's a custom prompt package bug, or something electron has done, but it'll be stuck on electron v30 until an update happens, or fork our own dialog package and fix the issue.

I'll go read up on the electron v31 release notes just to have a quick glance and see if something changed, but in case you do any work in the future involving electron v31+, that's just a head's up.

<br />

---

<br />

**Edit**: Spoke to the developer of custom prompts, he's pushing a hotfix, and submitting a bug report to Electron, it's apparently a bug from their v31 release.

custom prompts developer issued a hotfix which works. Electron will have to figure out a perm fix if they deem it an issue:
- https://github.com/electron/electron/issues/42975

Electron >= v31 now works fine.

**Comment by xdpirate** [comment]
Oh nice, great to hear. I tried to figure out an easier way to get a simple prompt before going the custom package route, initially trying with plain JS `prompt()`, but apparently it's not that easy, lol.

**Comment by Aetherinox** [comment]
Yeah, that's why I opted to reach out to the developer instead of going the manual way. I was shocked he replied so quickly, but he was right on it.

There's no reason to start investing large amounts of time for a simple dialog just for an electron wrapper.  It's doable, but it's a royal pain in the rear / time investment.

The only thing that would have been great out of box would have been a checkbox, I could add one, but not really a need.

Aside from doing some more testing, I've about done everything. I added the CLI arguments, but I also utilized that custom prompt interface to add a few settings via the menu (just for the lazies), so it can be done with either. Works good.  Right now I'm just letting it run while my server sends notifications to see how smooth it goes.

**Comment by Aetherinox** [comment]
Side question, do you have a preference or particular reason for not asar packing the app to cut down the resources folder?

I just did a test pack along with electron and then tested it on a remote linux and windows box, and everything works.

**Comment by xdpirate** [comment]
>asar packing 

This is my first and only Electron project so I have to admit I have no idea what that is. I learnt the basics of Electron specifically to make this app :) So I guess no - no reason.

**Comment by Aetherinox** [comment]
Ah ok, no worries, was just curious. Didn't want to start shipping it in another format if there was a reason. As simple as it is, it's a useful app. I hated keeping my browser open. I've had this app running for a few weeks now, and it has done great.

All asar packaging does is instead of having a resource folder with all the node modules and files scattered in the folder, you get a single .asar archive file that is used instead.

It's this
![thYwbYUNDS](https://github.com/user-attachments/assets/395aff7f-ab12-4dfe-86dd-a20fba612b7b)

Vs this:
![8xH4bREwL9](https://github.com/user-attachments/assets/1cf963f6-9084-44bb-8213-e0ea4e8480af)

Especially for the node_modules folder. And the other good thing is there's a "prune" command, which basically ensures all of the dev dependencies are removed from the final package, which cuts down on file size, since you don't need dev dependencies shipped with your product.

And that includes Electron, it doesn't need to be a dependency, it should be in the devDependencies list, as electron is already included when you pack the product, so by putting it in dependencies; you're basically shipping electron twice and increasing the size of the packaged product.

Just food for thought, Electron can be a pain in the rear at times, especially for a first product.

If you're curious as to the new command to pack as asar, you can look in the `package.json`, I added commands for Linux, mac, and Windows, and also created a `build.bat`, since I'm currently on a Windows machine (but I also tested this on Ubuntu 24 LTS.

Also, I snapped screenshots, if you want to see:
- https://github.com/Aetherinox/ntfy-desktop

I'm not sure what all you want included in the PR. I have a template I use when I develop new repos, which includes lint / pettier, and then a standard structure I use for READMEs. You can see it all in that repo.

Don't want to make it confusing for you, and you see all those files and think "Wtf do I do".

Still have some cleanup to do though, probably adding a production / dev env var for console, just because those make life easier to see what is going on.

**Comment by xdpirate** [comment]
Sorry, this totally fell off my radar as I got into the habit of opening the app whenever an audible notification arrived. If you're still interested in contributing, a PR that incorporates working native notifications and the minimize to tray functionality would be very welcome.
